src/main/java/com/treasuredata/client/TDClient.java

```diff
@@ -0,0 +1,527 @@
+
+        TDApiRequest.Builder request = TDApiRequest.Builder.POST(path);
+        for (Map.Entry<String, String> e : queryParam.entrySet()) {
+            request.addQueryParam(e.getKey(), e.getValue());
+        }
+        return httpClient.call(request.build(), apiKeyCache, resultTypeClass);
+    }
+
+    private String doPost(String path, Map<String, String> queryParam)
+            throws TDClientException
+    {
+        checkNotNull(path, "path is null");
+        checkNotNull(queryParam, "param is null");
+
+        TDApiRequest.Builder request = TDApiRequest.Builder.POST(path);
+        for (Map.Entry<String, String> e : queryParam.entrySet()) {
+            request.addQueryParam(e.getKey(), e.getValue());
+        }
+        return httpClient.call(request.build(), apiKeyCache);
+    }
+
+    private String doPost(String path)
+            throws TDClientException
+    {
+        TDApiRequest request = TDApiRequest.Builder.POST(path).build();
+        return httpClient.call(request, apiKeyCache);
+    }
+
+    private String doPut(String path, File filePath)
+            throws TDClientException
+    {
+        TDApiRequest request = TDApiRequest.Builder.PUT(path).setFile(filePath).build();
+        return httpClient.call(request, apiKeyCache);
+    }
+
+    /**
+     * Create a new TDClient that uses the given api key for the authentication.
+     * The new instance of TDClient shares the same HttpClient, so closing this will invalidate the other copy of TDClient instances
+     *
+     * @param newApiKey
+     * @return
+     */
+    @Override
+    public TDClient withApiKey(String newApiKey)
+    {
+        return new TDClient(config, httpClient, Optional.of(newApiKey));
+    }
+
+    @Override
+    public TDClient authenticate(String email, String password)
+    {
+        TDAuthenticationResult authResult = doPost("/v3/user/authenticate", ImmutableMap.of("user", email, "password", password), TDAuthenticationResult.class);
+        return withApiKey(authResult.getApikey());
+    }
+
+    @Override
+    public String serverStatus()
+    {
+        // No API key is requried for server_status
+        return httpClient.call(TDApiRequest.Builder.GET("/v3/system/server_status").build(), Optional.<String>absent());
+    }
+
+    @Override
+    public List<String> listDatabases()
+            throws TDClientException
+    {
+        TDDatabaseList result = doGet("/v3/database/list", TDDatabaseList.class);
+        List<String> tableList = new ArrayList<String>(result.getDatabases().size());
+        for (TDDatabase db : result.getDatabases()) {
+            tableList.add(db.getName());
+        }
+        return tableList;
+    }
+
+    private static Pattern databaseNamePattern = Pattern.compile("^([a-z0-9_]+)$");
+
+    public static String validateDatabaseName(String databaseName)
+    {
+        return validateName(databaseName, "Database");
+    }
+
+    private static String validateTableName(String tableName)
+    {
```

Comment:
Why is the modifier private? We should set same modifiers to validateDatabaseName and validateTableNamet methods.